### PR TITLE
lib/db: Don't panic getting a snapshot when db is closed (fixes #6319)

### LINKED ISSUE #6319: panic: database is closed when creating snapshot
Sentry Issue: [SYNCTHING-Y7](https://sentry.syncthing.net/syncthing/syncthing/issues/986/?referrer=github_integration)

```
Stacktrace (most recent call first):

  File "C:/BuildAgent/work/174e136266f8a219/lib/db/set.go", line 199, in Snapshot
    panic(err)
  File "C:/BuildAgent/work/174e136266f8a219/lib/model/folder.go", line 359, in scanSubdirs
    snap := f.fset.Snapshot()
  File "C:/BuildAgent/work/174e136266f8a219/lib/model/folder.go", line 583, in scanTimerFired
    err := f.scanSubdirs(nil)
  File "C:/BuildAgent/work/174e136266f8a219/lib/model/folder.go", line 173, in serve
    f.scanTimerFired()
  File "C:/BuildAgent/work/174e136266f8a219/lib/util/utils.go", line 183, in func1
    fn(ctx)
  File "C:/BuildAgent/work/174e136266f8a219/lib/util/utils.go", line 247, in Serve
    err = s.serve(s.ctx)
  File "C:/BuildAgent/work/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go", line 600, in func1

panic: database is closed
```

### PR DESCRIPTION
**Pull request by imsodin**
Regression introduced in #6239, thus RC material.

### CODE REVIEW

File: lib/db/set.go [hunk 1]
```diff
@@ -196,6 +196,9 @@ type Snapshot struct {
 func (s *FileSet) Snapshot() *Snapshot {
 	t, err := s.db.newReadOnlyTransaction()
 	if err != nil {
+		if backend.IsClosed(err) {
+			return nil
```
Comment (calmh) [inline lib/db/set.go]: I expect we're just going to panic one later when we call a method on the nil snapshot instead?

**Comment by calmh** [comment]
Nice rabbit hole you dug there, with the database closing stuff :D

File: lib/db/set.go (hunk 1 above)
Comment (imsodin) [reply lib/db/set.go]: No, we don't close the db until the model is stopped and the snapshot is only used there. It's the same as with `NewFileSet` (that rabbit hole wasn't dug by me for once as far as I remember :D ).

Comment (calmh) [reply lib/db/set.go]: Well, if this method is called at all then surely its return value will be used immediately thereafter?

Comment (imsodin) [reply lib/db/set.go]: Oh damn, so the panic shouldn't happen in the first place. I retract my statement and go back to thinking.

Comment (calmh) [reply lib/db/set.go]: And *cough* #5671 *cough*

**Comment by calmh** [comment]
Regardless, we need to move towards having error returns with clean exit paths for all database operations, so fixing this is necessary in the grand scheme of things.

**Comment by imsodin** [comment]
Agreed. And maybe this might not even be strictly RC material: The only way I can explain the panic (it's a single one so far) is that the model/folder didn't shut down in time so `syncthing.App` went ahead with the shutdown anyway, causing the panic. Same thing can happen with the file-set, which was already in 1.3.4. So we could just live with it for now and let it fall under https://github.com/syncthing/syncthing/issues/5907.

**Comment by calmh** [comment]
Hmm yes. If it's "just" a stray panic during exceptionally slow shutdown then I'm mostly OK with it. Lets let the RC simmer for a while and see how it looks.

**Pull request closed without merging**